aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,1479 @@
+                )
+            )
+        )) { return };
+
+        let amount = redeem_inactive_shares(pool, delegator_address, amount, withdrawal_lockup_epoch);
+        // short-circuit if amount to withdraw is 0 so no event is emitted
+        if (amount == 0) { return };
+
+        let stake_pool_signer = account::create_signer_with_capability(&pool.stake_pool_signer_cap);
+        stake::withdraw(&stake_pool_signer, amount);
+        coin::transfer<AptosCoin>(&stake_pool_signer, delegator_address, amount);
+
+        event::emit_event(
+            &mut pool.withdraw_stake_events,
+            WithdrawStakeEvent {
+                pool_address,
+                delegator_address,
+                amount_withdrawn: amount,
+            },
+        );
+    }
+
+    /// Return total stake owned by `delegator_address` within delegation pool `pool_address`
+    /// in each of its individual states.
+    public fun get_stake(pool_address: address, delegator_address: address): (u64, u64, u64) acquires DelegationPool {
+        assert_delegation_pool_exists(pool_address);
+        let pool = borrow_global<DelegationPool>(pool_address);
+
+        let (active, inactive, pending_active, pending_inactive) = stake::get_stake(pool_address);
+
+        active = pool_u64::shares_to_amount_with_total_coins(
+            &pool.active_shares,
+            pool_u64::shares(&pool.active_shares, delegator_address),
+            // use the most up-to-date total coins of the active share-pool
+            active + pending_active
+        );
+
+        // if no pending withdrawal, there is neither inactive nor pending_inactive stake
+        let (withdrawal_exists, withdrawal_lockup_epoch) = pending_withdrawal_exists(pool, delegator_address);
+        (inactive, pending_inactive) = if (withdrawal_exists) {
+            // delegator has either inactive or pending_inactive stake
+            let current_lockup_epoch = pool.current_lockup_epoch;
+            if (withdrawal_lockup_epoch < current_lockup_epoch) {
+                // if withdrawal's lockup epoch has been explicitly ended then its stake is certainly inactive
+                (
+                    pool_u64::balance(
+                        table::borrow(&pool.inactive_shares, withdrawal_lockup_epoch),
+                        delegator_address
+                    ),
+                    0
+                )
+            } else {
+                let pending_or_inactive_pool = table::borrow(&pool.inactive_shares, current_lockup_epoch);
+                let pending_or_inactive_shares = pool_u64::shares(pending_or_inactive_pool, delegator_address);
+                // if lockup already passed on the stake pool AND stake explicitly inactivated for validator,
+                // delegator's pending_inactive stake had also been inactivated on the delegation pool
+                if (last_reconfiguration_time() / MICRO_CONVERSION_FACTOR >= pool.locked_until_secs && pending_inactive == 0) {
```

Comment:
I think this could break when we introduce staking_contract as operator can interact directly with staking_contract to unlock commissions owned. This means we might not correctly advance observed lockup cycle here when real lockup cycle might have already advanced.